File: myLeetCode/Greedy/860. Lemonade Change.cpp
// 860. Lemonade Change
// https://leetcode.com/problems/lemonade-change/  

#include <iostream>
#include <vector>

bool lemonadeChange(std::vector<int>& bills) {
    int five = 0, ten = 0;

    for (int bill : bills) {
        if (bill == 5) {
            five++;
        } else if (bill == 10) {
            if (five == 0) return false;
            five--;
            ten++;
        } else {
            if (ten > 0 && five > 0) {
                ten--;
                five--;
            } else if (five >= 3) {
                five -= 3;
            } else {
                return false;
            }
        }
    }

    return true;
}

int main() {
    std::vector<int> bills1 = {5,5,5,10,20}; // Expected output:true
    std::cout << std::boolalpha << lemonadeChange(bills1) << std::endl;

    std::vector<int> bills2 = {5,5,10,10,20}; // Expected output:false
    std::cout << std::boolalpha << lemonadeChange(bills2) << std::endl;

    return 0;
}

File: C++/代码随想录/Greedy/860. Lemonade Change.py
def lemonadeChange(bills):
    five, ten = 0, 0
    for bill in bills:
        if bill == 5:
            five += 1
        elif bill == 10:
            if five == 0:
                return False
            five -= 1
            ten += 1
        else:
            if ten > 0 and five > 0:
                ten -= 1
                five -= 1
            elif five >= 3:
                five -= 3
            else:
                return False
    return True

# 测试用例
print(lemonadeChange([5, 5, 5, 10, 20]))  # True
print(lemonadeChange([5, 5, 10]))         # True
print(lemonadeChange([10, 10]))           # False
